app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/ReceiverDevice.kt

```diff
@@ -0,0 +1,111 @@
+import java.io.File
+import java.io.FileOutputStream
+import java.io.IOException
+import java.net.ServerSocket
+
+/**
+ * Helper class for the local file sharing module.
+ *
+ * Once the handshake has successfully taken place, this async-task is used to receive files from
+ * the sender device on the FILE_TRANSFER_PORT port. No. of files to be received (and their names)
+ * are learnt beforehand during the handshake.
+ *
+ * A single Task is used for the entire file transfer (the server socket accepts connections as
+ * many times as the no. of files).
+ */
+internal class ReceiverDevice(private val wifiDirectManager: WifiDirectManager) {
+  private var incomingFileName: String? = null
+  suspend fun receive() = withContext(Dispatchers.IO) {
+    try {
+      ServerSocket(WifiDirectManager.FILE_TRANSFER_PORT).use { serverSocket ->
+        Log.d(TAG, "Server: Socket opened at " + WifiDirectManager.FILE_TRANSFER_PORT)
+        val zimStorageRootPath = wifiDirectManager.zimStorageRootPath
+        val fileItems = wifiDirectManager.getFilesForTransfer()
+        var isTransferErrorFree = true
+        if (BuildConfig.DEBUG) Log.d(TAG, "Expecting " + fileItems.size + " files")
+        var fileItemIndex = 0
+        while (fileItemIndex < fileItems.size && this.isActive) {
+          incomingFileName = fileItems[fileItemIndex].fileName
+          try {
+            serverSocket.accept().use { client ->
+              if (BuildConfig.DEBUG) {
+                Log.d(TAG, "Sender device connected for " + fileItems[fileItemIndex].fileName)
+              }
+              publishProgress(fileItemIndex, FileItem.FileStatus.SENDING.ordinal)
+              val clientNoteFileLocation = File(zimStorageRootPath + incomingFileName)
+              val dirs = File(clientNoteFileLocation.parent)
+              if (!dirs.exists() && !dirs.mkdirs()) {
+                Log.d(TAG, "ERROR: Required parent directories couldn't be created")
+                isTransferErrorFree = false
+                fileItemIndex++
+              }
+              val fileCreated = clientNoteFileLocation.createNewFile()
+              if (BuildConfig.DEBUG) Log.d(TAG, "File creation: $fileCreated")
+              copyToOutputStream(
+                client.getInputStream(),
+                FileOutputStream(clientNoteFileLocation)
+              )
+              publishProgress(fileItemIndex, FileItem.FileStatus.SENT.ordinal)
+            }
+          } catch (e: IOException) {
+            Log.e(TAG, e.message)
+            isTransferErrorFree = false
+            publishProgress(fileItemIndex, FileItem.FileStatus.ERROR.ordinal)
+          }
+          fileItemIndex++
+        }
+        return@withContext isTransferErrorFree
+      }
+    } catch (e: IOException) {
+      Log.e(TAG, e.message)
+      return@withContext false // Returned when an error was encountered during transfer
+    }
+  }
+
+  private suspend fun publishProgress(vararg values: Int?) {
```

Comment:
this doesn't need to be varargs anymore, that was an asynctask limitation, just pass in the data we want

Reply:
we are calling this inside the recive()
` publishProgress(fileItemIndex, FileItem.FileStatus.ERROR.ordinal)`
How can we do that?

Reply:
pass in a fileIndex and a FileStatus I would think, we don't need these to be arrays anymore

Reply:
done